File: grid-math/project/wavelet-dist/haar-threaded-xform/myVector.m
//
//  myVector.m
//  oc-matrix
//
//  Created by Daniel D. Beatty on 4/4/05.
//  Copyright 2005 __MyCompanyName__. All rights reserved.
//

#import "myVector.h"


@implementation myVector
-(double) L2
{    
	int n = size;
    double temp = 0.0;
    int i;
    
    for ( i = 0; i < n; i++)
        temp += m_data[i] * m_data[i];
    return temp;
}

-(int) size
{
	return size;
}

-(void) setSize: (int) n
{
	size = n;
	[self allocate];
}

-(void) allocateWith: (int) i
{
	[self setSize:i];
	NSLog (@"Allocated a vector");
}

-(void) allocate
{
	[self deallocate];
	m_data = (double *) malloc(sizeof (double) * size);
}

-(void) deallocate 
{
	if ( m_data != NULL) free(m_data);
}


-(myVector *) copy
{
	int i, n;
    int m = size;
    //int o = x.Size();
    double temp;
    //if ( m != o)
    //    Size (o);
	myVector *aCopy = [[myVector init] alloc];
	[aCopy allocateWith:m];
    //m_data.reserve(o);
    //n = Size();
    for ( i = 0; i < m; i++)
    {
        //temp = x[i];
        //m_data[i] = temp;
		[aCopy vector][i] = m_data[i];
    }
    return (aCopy);
}

-(myVector *) addEqual: (myVector *) rhs
{
	
    int i ;
    int m = size;
    int o = [rhs size];
    if ( m != o)
        return self;
	
    for ( i = 0; i < m; i++)
        m_data[i] += [rhs vector][i];
    return self;
}

-(myVector *) subEqual: (myVector *) rhs
{
	
    int i ;
    int m = size;
    int o = [rhs size];
    if ( m != o)
        return self;
	
    for ( i = 0; i < m; i++)
        m_data[i] -= [rhs vector][i];
    return self;
}

-(myVector *) divEqual: (double) rhs
{
	
    int i ;
    int m = size;
	double drhs = 0;
	if ( rhs != 0) drhs= 1/ rhs;
	else return self;
	
    for ( i = 0; i < m; i++)
        m_data[i] *= drhs;
    return self;
}

-(myVector *) mulEqual: (double) rhs
{
	
    int i ;
    int m = size;
	
    for ( i = 0; i < m; i++)
        m_data[i] *= rhs;
    return self;
}


//-(myVector *) crossProduct: (myVector *) rhs;

-(double) dotProduct: (myVector *) rhs
{
    int n = [rhs size];
    double temp = 0.0;
    int i;
    
	
	if ( n != size ) return 0;
    for ( i = 0; i < n; i++)
        temp += m_data[i] * [rhs vector][i];
    return temp;
}

-(double *) vector
{
	return m_data;
}


@end

File: tags/original/wavelets/include/myVector.cpp
/*
 *  myVector.cpp
 *  numMethod6
 *
 *  Created by Daniel Beatty on Fri Nov 08 2002.
 *  Copyright (c) 2002 __MyCompanyName__. All rights reserved.
 *
 */

#include "myVector.h"
#include <stddef.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>

using namespace std;

double myVector::L2 ()
{
    double temp = dot();
    return temp;
}


myVector& myVector::operator= (myVector& x)
{
    int i, n;
    int m = Size();
    int o = x.Size();
    double temp;
    if ( m != o)
        Size (o);
    //m_data.reserve(o);
    n = Size();
    for ( i = 0; i < n; i++)
    {
        temp = x[i];
        m_data[i] = temp;
    }
    return (*this);
}

myVector& myVector::operator+= (/*const*/ myVector x)
{
    int i /*, n*/;
    int m = Size();
    int o = x.Size();
    if ( m != o)
        return (*this);
    //n = Size();
    for ( i = 0; i < m; i++)
        m_data[i] += x[i];
    return (*this);
}
myVector& myVector::operator-= (/*const*/ myVector x){
    int i /*, n*/;
    int m = Size();
    int o = x.Size();
    if ( m != o)
        return (*this);
    //n = Size();
    for ( i = 0; i < m; i++)
        m_data[i] -= x[i];
    return (*this);
}

myVector& myVector::operator*= (/*const*/ double x)
{
    int i, n;
    n = Size();
    for ( i = 0; i < n; i++)
        m_data[i] *= x;
    return (*this);
}

/*

myVector& myVector::mul ( matrixType A )

{
            int n = Size();
            int i;
            myVector y;
            matrixType temp(n,1);
            matrixType B;
            for (i = 0; i < n; i++)
                temp[i][0] = m_data[i];
            B = A; 
            B *= temp;
            for (i = 0; i < n; i++) {
                m_data[i] = y[i] = B[i][0];
            }
            return y;
}
*/
double myVector::dot ()
{
    int n = Size();
    double temp = 0.0;
    int i;
    double temp2;
    for ( i = 0; i < n; i++)
    {
    	temp2 = m_data[i];
        temp += pow ( temp2, double(2));
	}
    return temp;
}

double myVector::dot (/*const*/ myVector &x)
{
    int n = Size();
    double temp = 0.0;
    int i;
    
    for ( i = 0; i < n; i++)
        temp += m_data[i] * x[i];
    return temp;
}


myVector& myVector::operator/= (/*const*/ double x)
{
    int i, n;
    n = Size();
    for ( i = 0; i < n; i++)
        m_data[i] /= x;
    return (*this);
}

double myVector::sign (int i)
{
    double temp = 1;
        if ( m_data[i] < 0)
            temp = -1;
    return temp;
}

void myVector::Size (int n)
{
    //if ( m_data != NULL)
    //    delete m_data;
    m_data = new double[n];
    size = n;
}

void myVector::deallocate ()
{
	if (size > 0)
		//delete []m_data;
	if ( m_data != NULL) {
		delete []m_data;
			m_data = NULL;
		}
	size = 0;	
}

void myVector::allocate (int i)
{
	Size(i);
}
myVector::myVector (int Size)
{
    size = Size;
    m_data = new double[size];
    for ( int i = 0; i < size; i++)
        m_data[i] = 0.0;
}

myVector::myVector ()
{
    size = 0;
    m_data = NULL;
}

myVector::~myVector ()
{
	deallocate();
}
